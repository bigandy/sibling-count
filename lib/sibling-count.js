var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { LitElement, html } from "lit";
import { customElement } from "lit/decorators.js";
let SiblingCount = class SiblingCount extends LitElement {
    firstUpdated() {
        const slot = this.renderRoot.querySelector("slot");
        const parent = slot?.assignedNodes()[1];
        if (!parent) {
            return;
        }
        const siblingCount = parent.childElementCount;
        parent.style.setProperty("--sibling-count", siblingCount.toString());
        const siblings = parent.children;
        // Loop through all the children and add the custom property sibling-index to each.
        [...siblings].forEach((sibling, index) => {
            sibling.style.setProperty("--sibling-index", (index + 1).toString());
        });
    }
    render() {
        return html `<slot></slot>`;
    }
};
SiblingCount = __decorate([
    customElement("sibling-count")
], SiblingCount);
export { SiblingCount };
