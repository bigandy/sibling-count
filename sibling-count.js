const initialIndex = "initial-index";
const keepTrackOfUpdates = "keep-track-of-updates";

class SiblingCount extends HTMLElement {
  static observedAttributes = [initialIndex, keepTrackOfUpdates];

  constructor() {
    super();
    this.attachShadow({ mode: "open" });
    this.shadowRoot.innerHTML = `
	<slot></slot>
	`;
    this.modernAttrSupport = CSS.supports("x: attr(x type(*))");

    this.startingIndex = 1;
    this.keepTrackOfUpdates = false;
    // future friendly means that the user wants attr() to be used instead of inline custom properties.
    this.futureFriendly = false;

    if (this.hasAttribute(initialIndex)) {
      this.startingIndex = Number(this.getAttribute(initialIndex));
    }

    if (this.hasAttribute(keepTrackOfUpdates)) {
      this.keepTrackOfUpdates =
        this.getAttribute(keepTrackOfUpdates) === "true";
    }
  }

  setAttributeOrStyle(element, name, value) {
    if (this.modernAttrSupport && this.futureFriendly) {
      element.setAttribute(`data-${name}`, value);
    } else {
      element.style.setProperty(`--${name}`, value);
    }
  }

  showerAttributesOntoElements() {
    const shadow = this.shadowRoot;
    const slot = shadow.querySelector("slot");

    const assignedNodes = slot
      .assignedNodes()
      // filter out text nodes
      .filter((node) => node.nodeType === 1);

    const parent = assignedNodes[0];

    const siblingCount = parent.childElementCount;
    this.setAttributeOrStyle(parent, "sibling-count", siblingCount);

    const siblings = [...parent.children];
    // Loop through all the children and add the custom property sibling-index to each.

    siblings.forEach((sibling, index) => {
      const siblingIndex = index + this.startingIndex;
      this.setAttributeOrStyle(sibling, "sibling-index", siblingIndex);
    });
  }

  connectedCallback() {
    const shadow = this.shadowRoot;
    const slot = shadow.querySelector("slot");

    const filteredAssignedNodes = slot
      .assignedNodes()
      // filter out text nodes
      .filter((node) => node.nodeType === 1);

    // parent is the first filteredAssignedNodes.
    const [parent] = filteredAssignedNodes;

    this.showerAttributesOntoElements();

    if (this.keepTrackOfUpdates) {
      const observerCallback = (parentThis, mutations) => {
        mutations.forEach(() => parentThis.showerAttributesOntoElements());
      };

      // need to bind this so that the callback has access to it.
      const observer = new MutationObserver(observerCallback.bind(null, this));

      // configuration of the observer
      const config = { childList: true };

      observer.observe(parent, config);

      // disconnect the observer when the element is disconnected.
      this.addEventListener("element-disconnected", () => {
        observer.disconnect();
      });
    }
  }

  disconnectedCallback() {
    this.dispatchEvent(new Event("element-disconnected"));
  }
}

customElements.define("sibling-count", SiblingCount);
