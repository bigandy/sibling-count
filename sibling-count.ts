import { LitElement, html, type TemplateResult } from "lit";
import { customElement } from "lit/decorators.js";

@customElement("sibling-count")
export class SiblingCount extends LitElement {
  firstUpdated() {
    const slot = this.renderRoot.querySelector("slot");

    const parent = slot?.assignedNodes()[1] as HTMLElement;

    if (!parent) {
      return;
    }

    const siblingCount = parent.childElementCount;
    parent.style.setProperty("--sibling-count", siblingCount.toString());

    const siblings = parent.children;
    // Loop through all the children and add the custom property sibling-index to each.
    [...siblings].forEach((sibling, index) => {
      (sibling as HTMLElement).style.setProperty(
        "--sibling-index",
        (index + 1).toString()
      );
    });
  }

  render(): TemplateResult {
    return html`<slot></slot>`;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "sibling-count": SiblingCount;
  }
}
