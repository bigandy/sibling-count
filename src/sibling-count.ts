const keepTrackOfUpdates = "keep-track-of-updates";
const modernOnly = "modern-only";

interface SiblingCount {
  supportsSiblingCount: boolean;
  keepTrackOfUpdates: boolean;
  modernOnly: boolean;
}

class SiblingCount extends HTMLElement {
  static observedAttributes = [keepTrackOfUpdates, modernOnly];

  constructor() {
    super();
    this.attachShadow({ mode: "open" });

    if (!this.shadowRoot) {
      return;
    }
    this.shadowRoot.innerHTML = `
	<slot></slot>
	`;
    this.supportsSiblingCount =
      CSS?.supports("z-index: sibling-count()") ?? false;

    this.modernOnly = false;
    this.keepTrackOfUpdates = false;
  }

  setAttributeOrStyle(element: HTMLElement, name: string, value: string) {
    element.style.setProperty(`--${name}`, value);
  }

  showerAttributesOntoElements() {
    const shadow = this.shadowRoot;
    if (!shadow) {
      return;
    }
    const slot = shadow.querySelector("slot");

    const nonTextNodes =
      slot?.assignedNodes().filter(({ nodeType }) => nodeType === 1) ?? [];

    for (const node of nonTextNodes) {
      const parent = node as HTMLElement;

      const siblingCount = parent.childElementCount;
      if (siblingCount === 0) {
        console.warn(
          "Sibling Count - No children found. Use one parent element and this component will show how many children elements are present.",
        );
        return;
      }

      const siblings = [...parent.children];
      // Loop through all the children and add the custom property sibling-index to each.

      let index = 1;
      for (const sibling of siblings) {
        const siblingIndex = String(index);
        this.setAttributeOrStyle(
          sibling as HTMLElement,
          "sibling-index",
          siblingIndex,
        );

        this.setAttributeOrStyle(
          sibling as HTMLElement,
          "sibling-count",
          String(siblingCount),
        );

        index++;
      }
    }
  }

  connectedCallback() {
    if (this.hasAttribute(keepTrackOfUpdates)) {
      const attr = this.getAttribute(keepTrackOfUpdates) ?? "";
      this.keepTrackOfUpdates = ["true", ""].includes(attr);
    }

    if (this.hasAttribute(modernOnly)) {
      const attr = this.getAttribute(modernOnly) ?? "";
      this.modernOnly = ["true", ""].includes(attr);
    }

    if (this.supportsSiblingCount && this.modernOnly) {
      console.warn("Sibling Count - This browser supports sibling-count().");
      return;
    }

    const shadow = this.shadowRoot;
    if (!shadow) {
      return;
    }
    const slot = shadow.querySelector("slot");

    const filteredAssignedNodes = slot
      ?.assignedNodes()
      // filter out text nodes
      .filter((node) => node.nodeType === 1);

    // parent is the first filteredAssignedNodes.
    const parent = filteredAssignedNodes?.[0];

    if (!parent) {
      return;
    }

    this.showerAttributesOntoElements();

    if (this.keepTrackOfUpdates) {
      const observerCallback = (
        parentThis: SiblingCount,
        mutations: MutationRecord[],
      ) => {
        mutations.forEach(() => parentThis.showerAttributesOntoElements());
      };

      // need to bind this so that the callback has access to it.
      const observer = new MutationObserver(observerCallback.bind(null, this));

      // configuration of the observer
      const config = { childList: true };

      observer.observe(parent, config);

      // disconnect the observer when the element is disconnected.
      this.addEventListener("disconnected", () => {
        observer.disconnect();
      });
    }
  }

  disconnectedCallback() {
    if (this.keepTrackOfUpdates) {
      this.dispatchEvent(new Event("element-disconnected"));
    }
  }
}

customElements.define("sibling-count", SiblingCount);
export default SiblingCount;
