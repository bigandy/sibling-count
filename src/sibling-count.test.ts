import "./sibling-count.js";
import { afterEach, beforeEach, describe, it, expect } from "vitest";

describe("SiblingCount", () => {
  afterEach(() => {
    document.body.innerHTML = "";
    document.head.innerHTML = "";
  });

  beforeEach(() => {
    const siblingCount = document.createElement("sibling-count");
    siblingCount.innerHTML = `<ul>
                <li></li>
                <li></li>
                <li></li>
              </ul>
    `;
    document.body.appendChild(siblingCount);
  });

  it("should be in the document", () => {
    const siblingCount = document.querySelector("sibling-count");

    expect(siblingCount).toBeDefined();
  });

  it("the parent should have a --sibling-count matching the number of children", () => {
    const siblingCount = document.querySelector("sibling-count")!;

    const list = siblingCount.querySelector("ul")!;
    const listItems = list.querySelectorAll("li");
    const siblingCountValue =
      getComputedStyle(list).getPropertyValue("--sibling-count");

    expect(siblingCountValue).toBe("3");
    expect(listItems.length).toBe(3);
  });

  it("for each of the children: should have a --sibling-index matching the index + 1", () => {
    const siblingCount = document.querySelector("sibling-count")!;

    const list = siblingCount.querySelector("ul")!;
    const listItems = list.querySelectorAll("li");

    // Loop through list items, check their
    let index = 1;
    for (const listItem of listItems) {
      const siblingIndexValue =
        getComputedStyle(listItem).getPropertyValue("--sibling-index");
      expect(siblingIndexValue).toBe(String(index));
      index++;
    }
  });

  it("should match the snapshot", () => {
    const siblingCount = document.querySelector("sibling-count")!;
    expect(siblingCount).toMatchSnapshot();
  });
});
