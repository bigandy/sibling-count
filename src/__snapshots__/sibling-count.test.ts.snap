// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SiblingCount > should match the snapshot 1`] = `
<sibling-count>
  <ul
    style="--sibling-count: 3;"
  >
    
                
    <li
      style="--sibling-index: 1;"
    />
    
                
    <li
      style="--sibling-index: 2;"
    />
    
                
    <li
      style="--sibling-index: 3;"
    />
    
              
  </ul>
  
    
</sibling-count>
`;
